aptos-move/framework/aptos-stdlib/sources/table.move

```diff
@@ -96,6 +106,41 @@ module aptos_std::table {
     // ======================================================================================================
     // Internal API
 
+    struct NewTableEvent has drop, store {
+        handle: u128,
+        key_type: String,
+        value_type: String,
+    }
+
+    struct TableMetadata has key, store {
+        new_table_events: EventHandle<NewTableEvent>,
+    }
+
+    public fun initialize(account: &signer) {
+        assert!(address_of(account) == @aptos_std, error::permission_denied(ENOT_STD_ACCOUNT));
+
+        assert!(!exists<TableMetadata>(@aptos_std), error::already_exists(ETABLE_METADATA_ALREADY_EXISTS));
+        move_to(account, TableMetadata {
+            new_table_events: new_event_handle<NewTableEvent>(account)
+        });
+    }
+
+    /// Emits NewTableEvent if the event handle is installed under @aptos_std.
+    fun maybe_emit_create_table_event<K: copy + drop, V>(handle: u128) acquires TableMetadata {
+        // In unit tests the event handle is not installed.
+        if (exists<TableMetadata>(@aptos_std)) {
+            let key_type = type_name<K>();
+            let value_type = type_name<V>();
+
+            let table_metadata = borrow_global_mut<TableMetadata>(@aptos_std);
+            emit_event(&mut table_metadata.new_table_events, NewTableEvent {
```

Comment:
Wouldn't this disable parallel transaction execution if there are more than one transactions that create (different) tables?

Reply:
It will. We'll need to use an aggregator..

Do you see it happen @gelash ?

Reply:
aggregator only works for integer, probably not event handle

Reply:
the "count" in the handle, wouldn't it work?

Reply:
no, it won't. it requires a "read". aggregator works well in the case of delta add/sub but not read

Reply:
Oh.. true.. damn..

Reply:
Where can I find the implementation of how events are emitted? Based on this convo, I suppose there is a count that's read so we could add the new table event to TableMetaData? But how is TableMetaData stored (what data-structure), if it's being mutated too, then isn't that another conflict?

The key is to disperse the reads, always, so we should either develop more bucketed data-structures in the future (should make sense for event storage?), or find other crazy ways to aggregate things.

Reply:
More importantly: 
"type": "0x1::table::TableMetadata",
    "data": {
"new_table_events": {}

What else do we expect to put here? If this all goes to the same place, then adding anything else to TableMetaData, especially things that might be read often, will also cause those txns to have conflicts with arbitrary table creations, and opens a lot of vectors to slow down other txns by just creating tables. So we should be careful to not add more things, and if we don't, then we should probably just rename it.

Maybe we need a data-structure that uses a few buckets during the block to avoid conflicts but relies on executor to clear these brackets and prepare an aggregated representation at the end? Would be specific to append-able things, and would also need to deal with state-sync (but that we should probably break down into executing blocks at a time for other reasons anyway) - it's ugly, but maybe we can prettify, what do you think?

Reply:
I don't expect to put anything else there, just the event handle, which has a integer "counter" in there: https://github.com/aptos-labs/aptos-core/blob/09c8b36e7461ff6efa802ad13c7c83f266311c49/crates/aptos-rest-client/src/types.rs#L70

And the counter is bumped each time an event of that handle is emitted. https://github.com/aptos-labs/aptos-core/blob/3c2dae7c8f21dbc5a61bf0a1bbc1922cee8ec523/aptos-move/framework/aptos-stdlib/sources/event.move#L37

Making it 1000 handles instead of one will definitely help.. but it's so ugly. I'm exploring a different way to do it without events. "Sharding" the event handle will be a last sort that I don't expect to go with.

Reply:
If we're seeing that many table creations, we have introduced horrible paradigms and our storage will cry.

Reply:
> If we're seeing that many table creations

1 Million air drops in 10 minutes? (the gallery has a table in it)

Reply:
A BucketSmallTableThatAutomaticallyEvolvesToBucketTableTable would indeed be nice. @wrwg 
Though I heard we are not even using the BucketTable, because "the indexer can't index it"? @zekun000

Reply:
Table creations != Table insertions... both aren't great :trollface:

Reply:
I mean, a lot of people will initialize the gallery in short period of time, just to receive their first NFT.

Reply:
🤯

Reply:
> Making it 1000 handles instead of one will definitely help.. but it's so ugly. I'm exploring a different way to do it without events. "Sharding" the event handle will be a last sort that I don't expect to go with.
I agree, but just to make sure we are on the same page - much much smaller number than 1000 is enough for everything. Even if every txn was touching smt, we'd need smt like const * concurrency_level to basically have almost no conflicts, and probably concurrency_level many buckets would almost always make the amount of conflicts negligible, and I think we are currently running 4 threads?

Reply:
I agree back. From the storage device point of view we might need 32/64 threads concurrency to fully utilize the device, and then we can make sure each txn uses one of the event streams even if it creates multiple tables, by exposing the "txn hash" to Move. So probably 128 event handles will be good enough.

(and you are killing me by saying "smt" in place of "something" 😂 )